### DISCUSSION TITLE
Adding fonts at runtime

### DISCUSSION THREAD

**Discussion by leifniem** [khonsulabs/gooey#144] [Q&A] [answered]
Hi,

i was wondering if there is a way to load fonts at runtime. I tried figuring out a way for a few hours and ran into the limitation, that, as a far as i understand, i need access to the `Window` to modify the fonts in the list.

For context: I am trying to give an overview of fonts in a folder and their appearance and got the loading part in place and need a way to get them into the UI.

Thanks for all the great work, this framework feels quite mature for still being in development!

**Comment by ecton** [comment, marked as answer]
Apologies for not getting back to you yesterday, I have family visiting.

If you're just wanting to load the fonts at the start of the application, you can use `MakeWidget::into_window()` to convert any widget into a `Window`, which you can add the fonts to.

To add fonts after the window has been opened hasn't been designed yet, but I definitely want to support that as well. 

Thanks for trying out Cushy!

**Comment by leifniem** [comment]
Absolutely no apologies necessary :)

My use case would be the latter, since the program in question would be a font manager and therefore add and remove source folders containing fonts.

And happy to test Cushy! I have quite limited knowledge of Rust so far, but for not being a WebView-based UI-Framework this is quite accessible :)

**Reply by ecton** [reply]
There's probably a lot more missing functionality than this that you'll run into making a font manager, but I've implemented this in 3762bc6dc1925a0331e529d17364d17dc4d95602. [dynamic-fonts.rs](https://github.com/khonsulabs/cushy/blob/3762bc6dc1925a0331e529d17364d17dc4d95602/examples/dynamic-fonts.rs) is an example that accepts a path to a font and previews it.

Of note, cosmic-text's font fallback logic is currently fairly strict -- I spent a lot of time debugging this implementation only to discover the issue was that the two fonts I was testing had non-400 weight (which is the default weight in Cushy). Cosmic Text used a fallback in that situation. I've opened up an issue in their repository, but if you're having trouble getting a font to render, check that all of the face information matches what is set as the style for the widget.

Good luck!
